python/pyarrow/_flight.pyx

```diff
@@ -0,0 +1,486 @@
+
+        while True:
+            result = Result()
+            with nogil:
+                check_status(results.get().Next(&result.result))
+                if result.result == NULL:
+                    break
+            yield result
+
+    def list_flights(self):
+        """List the flights available on a service."""
+        cdef:
+            unique_ptr[CFlightListing] listing
+            FlightInfo result
+
+        with nogil:
+            check_status(self.client.get().ListFlights(&listing))
+
+        while True:
+            result = FlightInfo.__new__(FlightInfo)
+            with nogil:
+                check_status(listing.get().Next(&result.info))
+                if result.info == NULL:
+                    break
+            yield result
+
+    def get_flight_info(self, descriptor: FlightDescriptor):
+        """Request information about an available flight."""
+        cdef:
+            FlightInfo result = FlightInfo.__new__(FlightInfo)
+
+        with nogil:
+            check_status(self.client.get().GetFlightInfo(
+                descriptor.descriptor, &result.info))
+
+        return result
+
+    def do_get(self, ticket: Ticket, schema: Schema):
+        """Request the data for a flight."""
+        cdef:
+            # TODO: introduce unwrap
+            CTicket c_ticket
+            shared_ptr[CSchema] c_schema = pyarrow_unwrap_schema(schema)
+            unique_ptr[CRecordBatchReader] reader
+
+        c_ticket.ticket = ticket.ticket
+        with nogil:
+            check_status(self.client.get().DoGet(c_ticket, c_schema, &reader))
+        result = FlightRecordBatchReader()
+        result.reader.reset(reader.release())
+        return result
+
+    def do_put(self, descriptor: FlightDescriptor, schema: Schema):
+        """Upload data to a flight."""
+        cdef:
+            shared_ptr[CSchema] c_schema = pyarrow_unwrap_schema(schema)
+            unique_ptr[CRecordBatchWriter] writer
+
+        with nogil:
+            check_status(self.client.get().DoPut(
+                descriptor.descriptor, c_schema, &writer))
+        result = FlightRecordBatchWriter()
+        result.writer.reset(writer.release())
+        return result
+
+
+cdef class FlightDataStream:
+    cdef:
+        unique_ptr[CFlightDataStream] stream
+
+
+cdef class RecordBatchStream(FlightDataStream):
+    def __init__(self, reader):
+        # TODO: we don't really expose the readers in Python.
+        # self.stream.reset(None)
+        pass
+
+
+cdef void _get_flight_info(void* self, CFlightDescriptor c_descriptor,
+                           unique_ptr[CFlightInfo]* info):
+    """Callback for implementing Flight servers in Python."""
+    cdef FlightDescriptor descriptor = \
+        FlightDescriptor.__new__(FlightDescriptor)
+    descriptor.descriptor = c_descriptor
+    result = (<object> self).get_flight_info(descriptor)
+    if not result:
+        info.reset(NULL)
+        # TODO:
```

Comment:
Is there something missing here and below? I suppose you should convert `result` to a `CFlightInfo`?

Reply:
I haven't finished binding the server methods yet - I'll change this to raise NotImplementedError.